packages/itwin/tree-widget/src/components/trees/models-tree/ModelsTreeComponent.tsx

```diff
@@ -4,171 +4,255 @@
-  const showAll = useCallback(async () => {
-    await viewport?.addViewedModels(availableModels);
-    viewport?.clearNeverDrawn();
-    if (iModel) {
-      await toggleAllCategories(
-        IModelApp.viewManager,
-        iModel,
-        true,
-        viewport,
-        false,
-        undefined
-      );
-    }
-    viewport?.invalidateScene();
-  }, [viewport, availableModels, iModel]);
-
-  const viewToggle2D = useCallback(async () => {
-    if (is2dToggleActive) {
-      viewport?.changeModelDisplay(available2dModels, false);
-      setIs2dToggleActive(false);
-      setIcon2dToggle("icon-visibility-hide-2");
-    } else {
-      await viewport?.addViewedModels(available2dModels);
-      setIs2dToggleActive(true);
-      setIcon2dToggle("icon-visibility");
-    }
-    viewport?.invalidateScene();
-  }, [is2dToggleActive, viewport, available2dModels]);
-
-  const viewToggle3D = useCallback(async () => {
-    if (is3dToggleActive) {
-      viewport?.changeModelDisplay(available3dModels, false);
-      setIs3dToggleActive(false);
-      setIcon3dToggle("icon-visibility-hide-2");
-    } else {
-      await viewport?.addViewedModels(available3dModels);
-      setIs3dToggleActive(true);
-      setIcon3dToggle("icon-visibility");
-    }
-    viewport?.invalidateScene();
-  }, [is3dToggleActive, viewport, available3dModels]);
-
-  return (
-    <>
-      {iModel && viewport &&
-        <>
-          <TreeHeaderComponent
-            searchOptions={searchOptions}
-            showAll={showAll}
-            hideAll={hideAll}
-            invert={invert}
-            toggle2D={viewToggle2D}
-            toggle2DIcon={icon2dToggle}
-            toggle3D={viewToggle3D}
-            toggle3DIcon={icon3dToggle}
-          />
-          <AutoSizer>
-            {({ width, height }) => (
-              <ModelsTree
-                {...props}
-                iModel={iModel}
-                activeView={viewport}
-                width={width}
-                height={height}
-                filterInfo={{ filter: filterString, activeMatchIndex }}
-                onFilterApplied={onFilterApplied}
-              />
-            )}
-          </AutoSizer>
-        </>
+      await props.viewport.addViewedModels(notViewedModels);
+      props.viewport.changeModelDisplay(models, false);
+      props.viewport.invalidateScene();
+    };
+
+    return (
+      <IconButton
+        className="tree-widget-header-tree-toolbar-icon"
+        title={TreeWidget.translate("invert")}
+        icon="icon-visibility-invert"
+        onClick={invert}
+      />
+    );
+  }
+
+  export function View2DButton(props: ModelsTreeHeaderButtonProps) {
+    const models2d = useMemo(() => {
+      return props.models?.filter((model) => model.isPlanProjection).map((model) => model.id);
+    }, [props.models]);
+
+    const [is2dToggleActive, setIs2dToggleActive] = useState(() => (models2d && models2d.length !== 0) ? models2d.every((id) => props.viewport.viewsModel(id)) : false);
```

Comment:
Move logic determining if toggle is active to separate function to avoid duplicating it in `useState` and `useEffect`

Reply:
fixed